server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -548,6 +671,18 @@ private static String nodeIdFromIndex(int nodeIndex) {
             }
         }
 
+        void addNodesAndStabilise(int newNodesCount) {
+            addNodes(newNodesCount);
+            stabilise(
+                // The first pinging discovers the master
+                defaultMillis(DISCOVERY_FIND_PEERS_INTERVAL_SETTING)
+                    // One message delay to send a join
+                    + DEFAULT_DELAY_VARIABILITY
+                    // Commit a new cluster state with the new node(s). Might be split into multiple commits, and each might need a
+                    // followup reconfiguration
+                    + newNodesCount * 2 * DEFAULT_CLUSTER_STATE_UPDATE_DELAY);
```

Comment:
I wonder if we can find a stronger bound here when many nodes (i.e. more than 2) are joining.

There are two factors here:
- there is only a single pending reconfiguration in progress.
- all new nodes should have sent a join by `defaultMillis(DISCOVERY_FIND_PEERS_INTERVAL_SETTING) + DEFAULT_DELAY_VARIABILITY`
- joins are batched, so after `defaultMillis(DISCOVERY_FIND_PEERS_INTERVAL_SETTING) + DEFAULT_DELAY_VARIABILITY` i would expect for at most 4 cluster state updates to have to complete (an ongoing one, the follow-up reconfig, the one which batch-processes the remaining joins, and finally the reconfig for that one).

Reply:
I added TODOs in 9cfc15e6aa4fe3024879dd746e8111dade67738a and dbdaf0b26e23f2ebc11fa31438be6a16dfe1167c